### PR TITLE
fix(cli): Use urllib.parse for safe hostname extraction in error messages

### LINKED ISSUE #468: Use urllib.parse for hostname extraction in error messages
## Context

During review of PR #444, a security concern was identified: `format_network_error()` in `src/magpie/cli/errors.py` extracts hostnames via simple string splitting (`server.split("://", 1)[1].split("/", 1)[0]`). This is not robust for:
- URLs with embedded userinfo (e.g., `https://user:pass@host`)
- IPv6 addresses
- URLs with query strings or other components

If a user provides a server URL with credentials embedded (e.g., via `--server https://token:@artifacts.example.com`), the current implementation could inadvertently echo credentials back to the terminal in error messages.

## Proposed solution

Replace string splitting with `urllib.parse.urlsplit()` or `urllib.parse.urlparse()` in `format_network_error()`, and only display `hostname` and optionally `port`. Never include username, password, or query parameters in user-facing error messages.

## Example

```python
from urllib.parse import urlparse

def extract_hostname_safely(url: str) -> str:
    """Extract hostname (and optionally port) from URL, never credentials."""
    parsed = urlparse(url)
    if parsed.port:
        return f"{parsed.hostname}:{parsed.port}"
    return parsed.hostname or url
```

## References

- PR #444 review comment: https://github.com/SouthwestCCDC/magpie/pull/444#discussion_r2790598221
- Current implementation: `src/magpie/cli/errors.py` line 163-167

(AI-generated via Claude Code w/ Sonnet 4.5)

### PR DESCRIPTION
**Pull request by duplico**
## Summary

Fixes a security vulnerability where credentials embedded in server URLs could be leaked in CLI error messages.

Replaces insecure string splitting with `urllib.parse.urlparse()` to safely extract only the hostname and port from URLs, ensuring usernames and passwords are never displayed.

### Security Fix

- **Added `_extract_hostname_safely()` helper function** that uses `urllib.parse.urlparse()` to extract only hostname and port
- **Replaced vulnerable string splitting** in `format_network_error()`:
  - Before: `server.split("://", 1)[1].split("/", 1)[0]` (vulnerable to credential leaks)
  - After: `_extract_hostname_safely(server)` (safe, strips credentials)

### Test Coverage

Added two new tests to verify security:
- `test_credentials_not_leaked_in_error()` - Verifies credentials are stripped from URLs
- `test_credentials_with_port_not_leaked()` - Verifies credentials with custom ports are stripped

All existing tests continue to pass, including `test_hostname_extraction_from_url()` which validates that hostname:port extraction still works correctly.

## Security Impact

**Before this fix:**
```python
# User runs: magpie push --server https://token:secret@artifacts.example.com
# Error message shows: Could not connect to server 'token:secret@artifacts.example.com': DNS resolution failed
```

**After this fix:**
```python
# User runs: magpie push --server https://token:secret@artifacts.example.com
# Error message shows: Could not connect to server 'artifacts.example.com': DNS resolution failed
```

## Test Plan

- [x] Run unit tests: `uv run pytest tests/unit/ -v` (930 passed)
- [x] Lint checks pass: `uv run ruff check src/ tests/`
- [x] Format checks pass: `uv run ruff format src/ tests/`
- [x] Security tests verify credentials are not leaked
- [x] Existing hostname extraction tests still pass

## Related Issues

Fixes #468 (security follow-up from PR #444 review)

Generated with Claude Code w/ Sonnet 4.5

### CODE REVIEW

**Comment by duplico** [comment]
PR created for issue #468. Waiting for CI to complete.

(AI-generated via Claude Code w/ Sonnet 4.5)

**Comment by duplico** [comment]
CI passed. Waiting for reviews.

(AI-generated via Claude Code w/ Sonnet 4.5)

**Comment by duplico** [comment]
Reviews received. Starting Round 1: addressing 3 actionable comments.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/unit/test_cli_network_errors.py [hunk 1]
```diff
@@ -152,6 +152,39 @@ def test_hostname_extraction_from_url(self) -> None:
...
+            connect_error, server="https://user:secret_password@artifacts.example.com/api"
+        )
+
+        # Hostname should be present
+        assert "artifacts.example.com" in result
+        # Credentials should NOT be present
+        assert "user" not in result
+        assert "secret_password" not in result
+        # Protocol should not be in the hostname part
+        assert "https://" not in result
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/unit/test_cli_network_errors.py:209]: Fixed in 7f667c1. Updated test assertions to check for the full credential strings (user:secret_password@ and admin:pass123@) instead of individual words. This makes the tests more specific and robust against future message changes.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/magpie/cli/errors.py [hunk 2]
```diff
@@ -152,6 +153,40 @@ def handle_response_error(
...
+        parsed = urlparse(url)
+        if parsed.hostname:
+            # Include port if it's non-default
+            if parsed.port:
+                return f"{parsed.hostname}:{parsed.port}"
+            return parsed.hostname
+    except (ValueError, AttributeError):
+        # If parsing fails, return the original URL
+        pass
+    return url
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/cli/errors.py]: Fixed in 7f667c1. Updated the function to:

1. Wrap parsed.port access in try/except to catch ValueError for invalid ports
2. Return parsed.hostname even if port is invalid (instead of falling back to full URL)
3. Return safe placeholder "server" if parsing fails completely (never the original URL)

Added test_invalid_port_does_not_leak_credentials() to verify credentials are not leaked even with invalid ports like "https://user:pass@host:abc".

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/magpie/cli/errors.py [hunk 3]
```diff
@@ -152,6 +153,40 @@ def handle_response_error(
...
+        'artifacts.example.com'
+        >>> _extract_hostname_safely("https://artifacts.example.com")
+        'artifacts.example.com'
+    """
+    try:
+        parsed = urlparse(url)
+        if parsed.hostname:
+            # Include port if it's non-default
+            if parsed.port:
+                return f"{parsed.hostname}:{parsed.port}"
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/cli/errors.py]: Fixed in 7f667c1. Implemented the suggested IPv6 bracket notation:

- Detect ':' in hostname (indicates IPv6)
- Wrap in brackets when port is present
- Result: "[::1]:8443" instead of "::1:8443"

Added test_ipv6_with_port_formatted_correctly() to verify proper formatting.

(AI-generated via Claude Code w/ Sonnet 4.5)

**Comment by duplico** [comment]
## Round 1 Complete

**Comments addressed**: 3/3
**Commits**: 7f667c1

### Actions Taken
- Fixed test assertions to check full credential strings instead of individual words
- Added IPv6 bracket notation for addresses with ports ([::1]:8443)
- Secured parsing failure fallback to never return original URL
- Added graceful handling for invalid ports (catches ValueError)
- Added comprehensive tests for IPv6, invalid ports, and unparseable URLs

### Tests Added
- test_ipv6_with_port_formatted_correctly
- test_invalid_port_does_not_leak_credentials
- test_unparseable_url_uses_safe_placeholder

### Remaining
- 1 unresolved thread (waiting for confirmation)
- CI pending on latest commit
- Next: Wait for CI, then check if ready for adversarial review

(AI-generated via Claude Code w/ Sonnet 4.5)

**Comment by duplico** [comment]
## Adversarial Review: PASS

Performed security-focused review of the credential leak fix. No significant concerns found.

### Security Analysis

**Verified:**
- Credentials properly stripped via urlparse.hostname (excludes all userinfo)
- Safe fallback on parsing failures (returns 'server', never original URL)
- IPv6 addresses handled correctly with bracket notation
- Invalid ports handled gracefully (ValueError caught)

**Edge Cases Checked:**
- Empty hostname → guarded by 'if parsed.hostname:' check
- URL with only username → urlparse excludes ALL userinfo
- URL-encoded credentials → hostname excludes userinfo regardless
- Multiple @ symbols → urlparse treats first @ as separator
- Query params/fragments with secrets → hostname excludes these

**Test Coverage:**
- Credentials with password ✓
- Credentials with port ✓
- Invalid ports with credentials ✓
- Unparseable URLs ✓
- IPv6 with ports ✓

**Verdict:** Implementation is secure and ready for merge. Minor edge case tests (username-only, empty password) would be nice-to-have but are not blockers.

(AI-generated via Claude Code w/ Sonnet 4.5)

(AI-generated via Claude Code w/ Sonnet 4.5)

**Comment by duplico** [comment]
## Status: MERGE READY

All review comments have been addressed, CI passes, and adversarial security review found no concerns.

**Summary:**
- ✓ All 3 Copilot review comments addressed in commit 7f667c1
- ✓ CI passes (933 unit tests)
- ✓ Adversarial security review: PASS
- ✓ PR is mergeable

This PR is ready for merge.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/unit/test_cli_network_errors.py [hunk 4]
```diff
@@ -152,6 +152,78 @@ def test_hostname_extraction_from_url(self) -> None:
...
+        # Credentials should NOT be present
+        assert "user:pass@" not in result
+        assert "pass" not in result
+
+    def test_unparseable_url_uses_safe_placeholder(self) -> None:
+        """Completely unparseable URLs fall back to 'server' instead of leaking input."""
+        connect_error = httpx.ConnectError("Connection failed", request=MagicMock())
+
+        # Provide a malformed URL that might contain sensitive data
+        result = format_network_error(connect_error, server="not-a-url-with-secret-data")
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/unit/test_cli_network_errors.py]: Fixed in 4de3b45. Updated test to use a genuinely malformed URL with credentials ("://user:secret@host") instead of "not-a-url-with-secret-data" which parses as a path. This better demonstrates the security guarantee of returning 'server' placeholder instead of echoing back potentially sensitive input.

(AI-generated via Claude Code w/ Sonnet 4.5)
